src/Core/PluginCollection.php

```diff
@@ -0,0 +1,151 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright 2005-2011, Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.6.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Core;
+
+use ArrayIterator;
+use Cake\Core\Exception\MissingPluginException;
+use Countable;
+use InvalidArgumentException;
+use IteratorAggregate;
+use RuntimeException;
+
+/**
+ * Plugin Collection
+ *
+ * Holds onto plugin objects loaded into an application, and
+ * provides methods for iterating, and finding plugins based
+ * on criteria.
+ */
+class PluginCollection implements IteratorAggregate, Countable
+{
+    /**
+     * Plugin list
+     *
+     * @var array
+     */
+    protected $plugins = [];
+
+    const VALID_HOOKS = ['routes', 'bootstrap', 'console', 'middleware'];
```

Comment:
Should we add another valid hook into the array once #11692 is merged?

Reply:
Yes